File: leetcode/0109/0109.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    TreeNode* buildLevel(unsigned int level, std::list<TreeNode*>& nodes)
    {
        if(level == 0)
        {
            return nullptr;
        }

        if(level == 1)
        {
            auto* node = *(std::begin(nodes));
            nodes.pop_front();
            return node;
        }

        auto* left = buildLevel(level - 1, nodes);
        auto* root = *(std::begin(nodes));
        nodes.pop_front();
        auto* right = buildLevel(level - 1, nodes);
        root -> left = left;
        root -> right = right;

        return root;
    }

    TreeNode* sortedListToBST(ListNode* head)
    {
        unsigned int Size = 0;
        ListNode* Cursor = head;
        while(Cursor != nullptr)
        {
            Size++;
            Cursor = head -> next;
        }

        unsigned int i = 0;
        unsigned int Exp = 0;

    }
};

File: leetcode/0109/0109.py
#!/usr/bin/env python

import typing

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildLevel(self, size: int, nodes: typing.Iterator[TreeNode]) -> TreeNode:
        if size == 0:
            return None
        if size == 1:
            return next(nodes)

        HalfSize = size // 2
        Left = self.buildLevel(HalfSize, nodes)

        Root = next(nodes)

        Right = self.buildLevel(size - HalfSize - 1, nodes)

        Root.left = Left
        Root.right = Right
        return Root

    def sortedListToBST(self, head: ListNode) -> TreeNode:
        Size = 0
        Cursor = head

        while Cursor is not None:
            Size += 1
            Cursor = Cursor.next

        if Size == 0:
            return None

        Nodes = tuple(TreeNode(0) for i in range(Size))
        Cursor = head

        for Node in Nodes:
            Node.val = Cursor.val
            Cursor = Cursor.next

        Root = self.buildLevel(Size, iter(Nodes))
        return Root
